crates/oracle/src/oracle.rs

```diff
@@ -0,0 +1,332 @@
+                    } else {
+                        None
+                    }
+                })
+                .collect()
+        };
+
+        debug!(
+            messages = ?messages,
+            messages_count = messages.len(),
+            networks_count = available_networks.len(),
+            "Compressing message(s)."
+        );
+
+        let mut compression_engine = Encoder::new(CURRENT_ENCODING_VERSION, available_networks)
+            .expect(format!("Can't prepare for encoding because something went wrong",).as_str());
+        let encoded = compression_engine
+            .encode(&messages[..])
+            .expect(format!("Encoding failed: {:?}", messages).as_str());
+        debug!(encoded = ?encoded, "Successfully encoded message(s).");
+        Ok(encoded)
+    }
+
+    fn registered_networks(&self) -> Result<Vec<(Caip2ChainId, epoch_encoding::Network)>, Error> {
+        if self.subgraph_state.is_failed() {
+            todo!("Handle this as an error")
+        }
+        if self.subgraph_state.is_uninitialized() {
+            info!("Epoch Subgraph contains no initial state");
+            return Ok(Default::default());
+        };
+        info!("subgraph data is {:?}", self.subgraph_state.last_state());
+        Ok(self
+            .subgraph_state
+            .last_state()
+            .expect("expected data from a valid subgraph state, but found none")
+            .networks
+            .iter()
+            .map(|network| {
+                (
+                    network.id.clone(),
+                    epoch_encoding::Network {
+                        block_number: network.latest_block_number,
+                        block_delta: network.delta,
+                    },
+                )
+            })
+            .collect())
+    }
+}
+
+fn latest_blocks_to_message(latest_blocks: HashMap<Caip2ChainId, BlockPtr>) -> ee::Message {
+    Message::SetBlockNumbersForNextEpoch(
+        latest_blocks
+            .into_iter()
+            .map(|(chain_id, block_ptr)| (chain_id.as_str().to_owned(), block_ptr))
+            .collect(),
+    )
+}
+
+fn networks_diff_to_message(diff: &NetworksDiff) -> Option<ee::Message> {
+    if diff.deletions.is_empty() && diff.insertions.is_empty() {
+        None
+    } else {
+        Some(ee::Message::RegisterNetworks {
+            remove: diff.deletions.iter().map(|x| *x.1 as u64).collect(),
+            add: diff
+                .insertions
+                .iter()
+                .map(|x| x.0.as_str().to_string())
+                .collect(),
+        })
+    }
+}
+
+async fn submit_call<T>(
+    config: &Config,
+    protocol_chain: JrpcProviderForChain<T>,
+    payload: Vec<u8>,
+) -> web3::Result<H256>
+where
+    T: web3::Transport,
+{
+    let contract = Contract::from_json(
+        protocol_chain.web3.eth(),
+        config.contract_address,
+        include_bytes!("abi/data_edge.json"),
+    )
+    .expect("Can't read the ABI JSON file");
```

Comment:
Wouldn't it be more efficient if we compiled the contract only once?

Maybe we can store it in the new `Oracle` struct and pass it as a parameter to this function.